File: frontend/apps/main/src/api/device.ts
import http from './index'

export interface DeviceCheckUser {
  user_id: string
  display_name: string
  username: string | null
  family_name: string
  avatar_color: string
  role: string
  second_factor_type: string | null
  has_passkey: boolean
  last_seen_at: string
}

export interface DeviceCheckResponse {
  trusted: boolean
  users: DeviceCheckUser[]
}

export function checkDevice(deviceId: string) {
  return http.post<DeviceCheckResponse>('/auth/device/check', { device_id: deviceId })
}

export interface DeviceSelectResponse {
  second_factor_required: boolean
  temp_token?: string
  second_factor_type?: string
  display_name?: string
  avatar_color?: string
}

export function selectDeviceUser(deviceId: string, userId: string, altcha?: string) {
  return http.post<DeviceSelectResponse>('/auth/device/select', {
    device_id: deviceId,
    user_id: userId,
    altcha,
  })
}

export interface DeviceTrustResponse {
  session_id: string
  device_id: string
  device_name: string
  expires_at: string
}

export interface DeviceSession {
  session_id: string
  device_id: string | null
  device_name: string
  created_at: string
  last_seen_at: string
  expires_at: string
  is_current: boolean
}

export function listDevices() {
  return http.get<DeviceSession[]>('/auth/devices')
}

export function revokeDevice(sessionId: string) {
  return http.delete(`/auth/devices/${sessionId}`)
}

export function revokeAllDevices() {
  return http.delete('/auth/devices')
}

export interface FamilyDevice {
  session_id: string
  device_id: string | null
  user_id: string
  display_name: string
  avatar_color: string
  device_name: string
  last_seen_at: string
  created_at: string
  is_current: boolean
}

export async function listFamilyDevices() {
  const { data } = await http.get<FamilyDevice[]>('/auth/devices/family')
  return data
}

// --- WebAuthn Device Authentication ---

export interface WebAuthnAuthOptionsResponse {
  options: Record<string, unknown>
  challenge: string
}

export function getDeviceWebAuthnAuthOptions(deviceId: string, userId: string) {
  return http.post<WebAuthnAuthOptionsResponse>('/auth/device/webauthn/auth-options', {
    device_id: deviceId,
    user_id: userId,
  })
}

export function verifyDeviceWebAuthn(
  deviceId: string,
  userId: string,
  credential: Record<string, unknown>,
  challenge: string,
) {
  return http.post<DeviceSelectResponse>('/auth/device/webauthn/verify', {
    device_id: deviceId,
    user_id: userId,
    credential,
    challenge,
  })
}

// --- WebAuthn Device Trust Registration ---

export interface WebAuthnRegisterOptionsResponse {
  options: Record<string, unknown>
  challenge: string
}

export function getDeviceTrustWebAuthnOptions() {
  return http.post<WebAuthnRegisterOptionsResponse>('/auth/device/trust/webauthn/register-options')
}

export function registerDeviceTrustWebAuthn(
  credential: Record<string, unknown>,
  challenge: string,
) {
  return http.post<{ registered: boolean }>('/auth/device/trust/webauthn/register', {
    credential,
    challenge,
  })
}

File: server/apps/backend/app/schemas/device.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, field_validator

from apps.backend.app.schemas.base import SnowflakeBase


class DeviceTrustResponse(SnowflakeBase):
    session_id: int
    device_id: str
    device_name: str
    expires_at: datetime


class DeviceSessionResponse(SnowflakeBase):
    session_id: int
    device_id: str | None
    device_name: str
    created_at: datetime
    last_seen_at: datetime
    expires_at: datetime
    is_current: bool


class DeviceCheckRequest(BaseModel):
    device_id: str


class DeviceCheckUserItem(SnowflakeBase):
    user_id: int
    display_name: str
    username: str | None
    family_name: str
    avatar_color: str
    role: str
    second_factor_type: str | None
    has_passkey: bool
    last_seen_at: datetime


class DeviceCheckResponse(BaseModel):
    trusted: bool
    users: list[DeviceCheckUserItem] = []


class DeviceSelectRequest(BaseModel):
    device_id: str
    user_id: str
    altcha: str | None = None  # Optional; only verified in production

    @field_validator("user_id")
    @classmethod
    def validate_user_id(cls, v: str) -> str:
        if not v.isdigit() or len(v) > 20:
            raise ValueError("invalid user_id")
        return v


class DeviceSelectResponse(BaseModel):
    second_factor_required: bool
    temp_token: str | None = None
    second_factor_type: str | None = None
    display_name: str | None = None
    avatar_color: str | None = None


class DeviceTrustRequest(BaseModel):
    device_id: str | None = None


class FamilyDeviceResponse(SnowflakeBase):
    session_id: int
    device_id: str | None
    user_id: int
    display_name: str
    avatar_color: str
    device_name: str
    last_seen_at: datetime
    created_at: datetime
    is_current: bool


class DeviceWebAuthnAuthOptionsRequest(BaseModel):
    """Request challenge for WebAuthn device authentication."""
    device_id: str
    user_id: str

    @field_validator("user_id")
    @classmethod
    def validate_user_id(cls, v: str) -> str:
        if not v.isdigit() or len(v) > 20:
            raise ValueError("invalid user_id")
        return v


class DeviceWebAuthnAuthOptionsResponse(BaseModel):
    """WebAuthn authentication challenge."""
    options: dict[str, Any]
    challenge: str


class DeviceWebAuthnVerifyRequest(BaseModel):
    """Submit WebAuthn authentication response."""
    device_id: str
    user_id: str
    credential: dict[str, Any]
    challenge: str

    @field_validator("user_id")
    @classmethod
    def validate_user_id(cls, v: str) -> str:
        if not v.isdigit() or len(v) > 20:
            raise ValueError("invalid user_id")
        return v


class DeviceWebAuthnVerifyResponse(BaseModel):
    """Result after WebAuthn verification — same shape as DeviceSelectResponse."""
    second_factor_required: bool
    temp_token: str | None = None
    second_factor_type: str | None = None
    display_name: str | None = None
    avatar_color: str | None = None


class DeviceTrustWebAuthnOptionsResponse(BaseModel):
    """Registration options for WebAuthn during device trust."""
    options: dict[str, Any]
    challenge: str


class DeviceTrustWebAuthnRegisterRequest(BaseModel):
    """Complete WebAuthn registration during device trust."""
    credential: dict[str, Any]
    challenge: str
